packages/go_router/lib/src/delegate.dart

```diff
@@ -44,13 +44,21 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
 
   final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
 
+  /// The list of completers for the promises when pushing asynchronous routes.
+  final Map<String, Completer<dynamic>> completers =
+      <String, Completer<dynamic>>{};
   RouteMatchList _matches = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
 
-  /// Pushes the given location onto the page stack
-  void push(RouteMatch match) {
-    // Remap the pageKey to allow any number of the same page on the stack
+  /// Pushes the given location onto the page stack with an optional promise.
+  Future<T?> push<T extends Object?>(RouteMatch match) {
+    // Remap the pageKey to allow any number of the same page on the stack.
     final String fullPath = match.fullpath;
+
+    // Create a completer for the promise and store it in the completers map.
+    final Completer<T?> completer = Completer<T?>();
+    completers[fullPath] = completer;
```

Comment:
This will break when if you push the same screen more than once, since this will overwrite the old completer and to the user it will never get cancelled. You need to do the same thing as `pageKey` below but ideally these would be stored on the route match (see my comment above)